### ISSUE TITLE
Publish documentation for C# compiler features not covered in the C# language spec

### ISSUE THREAD

**Issue by VladimirReshetnikov** [dotnet/roslyn#411]
Publish documentation for C# compiler features not covered in the C# language spec:
- command line format
- COM-specific features
- named indexers
- special attributes
- limitations (e.g. identifier length, fully qualified name length, ...)
- etc

**Comment by AdamSpeight2008** [comment]
On the language specifications. Does the licencing permit? Would it allow me to print a personal copy of both C# and VB.net specifications, say from a commercial printing company

**Comment by gafter** [comment]
@AdamSpeight2008 I don't think we can provide legal advice here.

**Comment by VladimirReshetnikov** [comment]
This task is closely related (and depends on) #515.

**Comment by ufcpp** [comment]
@VladimirReshetnikov Some Japanese developers were aware that C# 6.0 has a breaking chage related to the Unicode character classes: Katakana Middle Dot is no longer available for identifiers in C# 6.0.
https://github.com/ufcpp/UfcppSample/blob/master/BreakingChanges/VS2015_CS6/KatakanaMiddleDot.cs
This change would be better to be documented.What do you think?

**Comment by gafter** [comment]
@ufcpp Noted in #1070

**Comment by ufcpp** [comment]
@gafter thanks!

**Comment by jcouv** [comment]
Here's one: `#error version` will print a diagnostic with the version of the compiler and the langversion setting.
More details in https://github.com/dotnet/roslyn/pull/18045

**Comment by jcouv** [comment]
The command-line options are now documented:
- [csc.exe command-line options](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/compiler-options/)
- [vbc.exe command-line options](https://docs.microsoft.com/en-us/dotnet/visual-basic/reference/command-line-compiler/)

**Comment by Youssef1313** [comment]
@BillWagner Is anything missing?

**Comment by BillWagner** [comment]
@Youssef1313 

I'm confident we have all the command line options. I went through the list with @jaredpar and @chsienki when I built the new pages. 

(If anything is missing, I'd rather move the issue to [dotnet/docs](https://github.com/dotnet/docs) so we can track it.

**Comment by Youssef1313** [comment]
@BillWagner There are more items in the original issue:

> COM-specific features
> named indexers
> special attributes
> limitations (e.g. identifier length, fully qualified name length, ...)
> etc

**Comment by BillWagner** [comment]
Thanks @Youssef1313 

On each of these:

- **COM-specific features**:  Most everything we have is [here](https://docs.microsoft.com/dotnet/csharp/programming-guide/interop/)  Is there something specific that's missing?
- **named indexers**: Is this meant to refer to indexed properties? If so, consuming them is covered [here](https://docs.microsoft.com/dotnet/csharp/programming-guide/interop/how-to-use-indexed-properties-in-com-interop-rogramming). Indexed properties can't be created in C#.
- **special attributes**: [This section](https://docs.microsoft.com/dotnet/csharp/language-reference/attributes/global) covers all those that are supported. The compiler does add other attributes in metadata for a number of features, but those are only meant to be read by the compiler. We don't write those.
- **limitations (e.g. identifier length, fully qualified name length, ...)** We haven't covered those. I looked in the standard, and these limits are mandated there. I'm curious as to the real need here. Do you know of a reason for this request?

**Comment by Youssef1313** [comment]
> limitations (e.g. identifier length, fully qualified name length, ...) We haven't covered those. I looked in the standard, and these limits are mandated there. I'm curious as to the real need here. Do you know of a reason for this request?

Issue title mentions "not covered in the C# language spec". Given that they're stated in the standard, I think there is no need to doc them.

I think the issue can be closed given that everything is covered.

**Comment by gafter** [comment]
I don't think that using named indexers (defined in other languages) is specified.

There is a fairly substantial OneNote notebook that I left for the compiler team (called something like "Compiler Specification") that describes many issues worth documenting.  It is possible that all of the things in that have been documented (or apparently are not worth documenting).  See https://github.com/dotnet/roslyn/pull/536 for the start of the transition from that document to public documents.

**Comment by gafter** [comment]
@Youssef1313 Where in the spec would I find the limit for the length of an identifier?  I looked in https://github.com/dotnet/csharpstandard/blob/draft-v6/standard/lexical-structure.md#643-identifiers but did not notice it.

**Comment by Youssef1313** [comment]
> Where in the spec would I find the limit for the length of an identifier?

Yeah it seems it's not there. Maybe I misunderstood this part of @BillWagner's comment (`I looked in the standard, and these limits are mandated there.`).

**Comment by BillWagner** [comment]
@Youssef1313 @gafter That's my typo above. I also looked in the standard in the lexical structure, and in the variables clause. Identifier limits aren't mentioned in the standard. They should be.

**Comment by JoeRobich** [comment]
Reopening as this was accidentally closed when we merged the history from the roslyn-sdk repo into roslyn.
